File: src/jmgo_controller/remote.py
from __future__ import annotations

import json
import socket
import time
from collections.abc import Iterator
from typing import Any

from .protocol import (
    ProtocolError,
    decode_state,
    key_packet,
    redact_state,
    sanitize_state,
    set_volume_packet,
)

DEFAULT_PORT = 9005

KEY_CODES = {
    "back": 4,
    "up": 19,
    "down": 20,
    "left": 21,
    "right": 22,
    "ok": 23,
    "volume-up": 24,
    "volume-down": 25,
    "menu": 82,
    "settings": 605,
    "home": 706,
    "power": 707,
    "power-menu": 2011,
}


class Remote:
    def __init__(self, host: str, port: int = DEFAULT_PORT, timeout: float = 3.0):
        self.host = host
        self.port = port
        self.timeout = timeout

    def _connect(self) -> socket.socket:
        return socket.create_connection((self.host, self.port), timeout=self.timeout)

    def press(self, key: str) -> None:
        try:
            keycode = KEY_CODES[key]
        except KeyError as error:
            raise ValueError(f"unknown key: {key}") from error
        with self._connect() as connection:
            connection.sendall(key_packet(keycode, True))
            time.sleep(0.12)
            connection.sendall(key_packet(keycode, False))

    def set_volume(self, volume: int) -> None:
        with self._connect() as connection:
            connection.sendall(set_volume_packet(volume))

    def read_state(self, wait: float = 1.0, include_identifiers: bool = False) -> dict[str, Any]:
        chunks: list[bytes] = []
        with self._connect() as connection:
            connection.settimeout(wait)
            while True:
                try:
                    chunk = connection.recv(65536)
                except TimeoutError:
                    break
                if not chunk:
                    break
                chunks.append(chunk)
        state = decode_state(b"".join(chunks))
        return sanitize_state(state) if include_identifiers else redact_state(state)

    def watch(self, include_identifiers: bool = False) -> Iterator[dict[str, Any]]:
        buffer = bytearray()
        with self._connect() as connection:
            connection.settimeout(None)
            while chunk := connection.recv(65536):
                buffer.extend(chunk)
                try:
                    state = decode_state(bytes(buffer))
                except ProtocolError:
                    continue
                if state:
                    yield sanitize_state(state) if include_identifiers else redact_state(state)
                    buffer.clear()

    def print_watch(self, include_identifiers: bool = False) -> None:
        for state in self.watch(include_identifiers):
            print(json.dumps(state, ensure_ascii=False, sort_keys=True), flush=True)

File: src/remote.ts
import { createConnection, type Socket } from "node:net";
import {
  decodeState,
  frame,
  keyPacket,
  redactState,
  sanitizeState,
  setVolumePacket,
  splitFrames,
  type ProjectorState,
} from "./protocol.js";

export const keyCodes = {
  back: 4,
  up: 19,
  down: 20,
  left: 21,
  right: 22,
  ok: 23,
  "volume-up": 24,
  "volume-down": 25,
  menu: 82,
  settings: 605,
  home: 706,
  power: 707,
  "power-menu": 2011,
} as const;

export type RemoteKey = keyof typeof keyCodes;

function delay(milliseconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

export class Remote {
  constructor(
    readonly host: string,
    readonly port = 9005,
    readonly timeoutMs = 3_000,
  ) {}

  private connect(): Promise<Socket> {
    return new Promise((resolve, reject) => {
      const socket = createConnection({ host: this.host, port: this.port });
      const onError = (error: Error) => reject(error);
      socket.once("error", onError);
      socket.once("connect", () => {
        socket.off("error", onError);
        resolve(socket);
      });
      socket.setTimeout(this.timeoutMs, () => socket.destroy(new Error("connection timed out")));
    });
  }

  async press(key: RemoteKey): Promise<void> {
    const socket = await this.connect();
    try {
      socket.write(keyPacket(keyCodes[key], true));
      await delay(120);
      await new Promise<void>((resolve, reject) => {
        socket.once("error", reject);
        socket.end(keyPacket(keyCodes[key], false), resolve);
      });
    } finally {
      socket.destroy();
    }
  }

  async setVolume(volume: number): Promise<void> {
    const socket = await this.connect();
    await new Promise<void>((resolve, reject) => {
      socket.once("error", reject);
      socket.end(setVolumePacket(volume), resolve);
    });
  }

  async readState(waitMs = 1_000, includeIdentifiers = false): Promise<ProjectorState> {
    const socket = await this.connect();
    const chunks: Buffer[] = [];
    const data = await new Promise<Buffer>((resolve, reject) => {
      let settled = false;
      const finish = () => {
        if (settled) return;
        settled = true;
        socket.destroy();
        resolve(Buffer.concat(chunks));
      };
      socket.on("data", (chunk: Buffer) => chunks.push(chunk));
      socket.once("end", finish);
      socket.once("error", reject);
      socket.setTimeout(waitMs, finish);
    });
    const state = decodeState(data);
    return (includeIdentifiers ? sanitizeState(state) : redactState(state)) as ProjectorState;
  }

  async *watch(includeIdentifiers = false): AsyncGenerator<ProjectorState> {
    const socket = await this.connect();
    socket.setTimeout(0);
    let buffer = Buffer.alloc(0);
    try {
      for await (const chunk of socket) {
        buffer = Buffer.concat([buffer, chunk as Buffer]);
        const split = splitFrames(buffer);
        buffer = Buffer.from(split.remainder);
        for (const payload of split.frames) {
          const framed = frame(payload);
          const state = decodeState(framed);
          yield (includeIdentifiers ? sanitizeState(state) : redactState(state)) as ProjectorState;
        }
      }
    } finally {
      socket.destroy();
    }
  }
}
